### DISCUSSION TITLE
Code Modernizing / Config, does not behave as expected

### DISCUSSION THREAD

**Discussion by ItsReddi** [rectorphp/rector#8823] [Q&A] [answered]
Assuming the following config:
```php
->withPhpSets(
        //Max Optimized PHP Version, with all new features/optimizations enabled, but limited to withPhpVersion features
        php84: true 
    )
    //Max PHP Verision we want to be compatible with, rector will not use any features above this version
    ->withPhpVersion(PhpVersion::PHP_81);
```

I would expect:
https://github.com/rectorphp/rector/blob/main/rules/Php84/Rector/Param/ExplicitNullableParamTypeRector.php
is running, since it is not a PHP 8.4 syntax and would work on PHP 8.1.

Do i misunderstand the configurations? (see comments)

Thank you

**Comment by samsonasik** [comment, marked as answer]
For that, it is better to you need to use php 8.4 runtime, otherwise, it may cause unexpected error, eg on required after optional, see https://3v4l.org/5jms6, without nullable, it is working https://3v4l.org/4O5MK, that's why upgrade php step by step is better, to ensure per-feature applied correctly in order.

But if you want it so hard, you can do something like:

```php
<?php

return \Rector\Config\RectorConfig::configure()
    ->withPaths([__DIR__ . '/src'])
    ->withPhpSets(php83: true)
    ->withPhpVersion(\Rector\ValueObject\PhpVersion::PHP_84)
    ->withRules([
         \Rector\Php84\Rector\Param\ExplicitNullableParamTypeRector::class,
    ]);
```

with the risk is on yours, see answer at issue https://github.com/rectorphp/rector/issues/8761#issuecomment-2257692788

**Reply by ItsReddi** [reply]
Thanks. https://3v4l.org/5jms6, explains it. But i am curios, wouldn't it also cause an unexpected error on 8.4?


Is my general assumption about the above config options is correct?
Got my assumption from Docs & https://github.com/rectorphp/rector/discussions/7520#discussioncomment-3952765

**Reply by samsonasik** [reply]
Yes, but on php 8.0, we have rule for it https://getrector.com/rule-detail/optional-parameters-after-required-rector

While nullable param exists since php 7.1, the upgrading should be by order, otherwise, it will got that error too early, that's why it is better to upgrade version by version, and verify, so you have safe result.

**Reply by ItsReddi** [reply]
Thank you for your patience.

To clarify, I am fine with this specific rule not being executed, but I still don't understand why. I'm already using PHP 8.1, so the 8.0 rules should work, and the 8.4 rules should be executed safely. Since I don't fully grasp this, I am unsure of the correct settings.

My goals:
- Code must be compatible with PHP 8.1
- I want to modernize the code to address deprecations and breaking changes in PHP 8.2, 8.3, and 8.4, which can already be fixed in PHP 8.1.

Here's my configuration:
```php
->withPhpSets(php84: true)
->withPhpVersion(PhpVersion::PHP_81);
```

Or should it be:
```php
->withPhpSets(php81: true)
->withPhpVersion(PhpVersion::PHP_84);
```

**Reply by samsonasik** [reply]
You can do:

```php
->withPhpSets(php81: true)
->withPhpVersion(PhpVersion::PHP_84)
->withRules([
         \Rector\Php84\Rector\Param\ExplicitNullableParamTypeRector::class,
    ]);
```

**Reply by ItsReddi** [reply]
Thanks.

After experimenting with a deprecation from PHP 8.2, I realized I might have misunderstood how the tool works. It looks like I need to explicitly set the rules that can be applied. If i want to early address depreciations and breaking changes.
Correct?

I initially assumed that:

```php
->withPhpSets(php84: true)
->withPhpVersion(PhpVersion::PHP_81);
```

would work automatically like this:

```php
->withPhpSets(php81: true)
->withPhpVersion(PhpVersion::PHP_84)
->withRules([
    ExplicitNullableParamTypeRector::class,
    VariableInStringInterpolationFixerRector::class,
    // other deprecations fixable in 8.1
]);
```

But that assumption seems entirely incorrect, right?

**Reply by samsonasik** [reply]
`withPhpSets()` is collecting rules up until defined, while `withPhpVersion` enforce php version, only when php version and rules match, it will be executed, see `provideMinPhpversion()` method:

https://github.com/rectorphp/rector-src/blob/71fb2e4c9346bf6ce6b057ff3d77511662f7913c/rules/Php84/Rector/Param/ExplicitNullableParamTypeRector.php#L103-L106
